### PR TITLE
Remove plugin interfaces

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

File: src/neo/Utility.cs [hunk 1]
```diff
@@ -29,6 +30,8 @@ public Logger()
             }
         }
 
+        public static event LogEventHandler Logging;
```
Comment (shargon) [inline src/neo/Utility.cs:33]: `OnLog`?

File: src/neo/NeoSystem.cs [hunk 2]
```diff
@@ -119,7 +119,7 @@ public NeoSystem(ProtocolSettings settings, string storageEngine = null, string
         {
             this.Settings = settings;
             this.GenesisBlock = CreateGenesisBlock(settings);
-            this.storage_engine = storageEngine;
+            this.storage_engine = storageEngine ?? nameof(MemoryStore);
```
Comment (shargon) [inline src/neo/NeoSystem.cs:122]: Maybe we should log that the memory store is used

File: src/neo/Ledger/MemoryPool.cs [hunk 3]
```diff
@@ -306,12 +308,14 @@ internal VerifyResult TryAdd(Transaction tx, DataCache snapshot)
...
-                    plugin.TransactionsRemoved(_system, MemoryPoolTxRemovalReason.CapacityExceeded, removedTransactions);
-            }
+            TransactionAdded?.Invoke(this, poolItem.Tx);
+            if (removedTransactions != null)
+                foreach (Transaction removed in removedTransactions)
+                    TransactionRemoved?.Invoke(this, new()
+                    {
+                        Transaction = removed,
+                        Reason = TransactionRemovalReason.CapacityExceeded
+                    });
```
Comment (shargon) [inline src/neo/Ledger/MemoryPool.cs:318]: Use the array in the event?, if the plugin want to iterate, it should be his responsibility

```suggestion
                  TransactionRemoved?.Invoke(this, new()
                  {
                      Transactions = removedTransactions,
                      Reason = TransactionRemovalReason.CapacityExceeded
                  });
```

File: src/neo/Ledger/MemoryPool.cs [hunk 4]
```diff
@@ -476,8 +480,12 @@ internal void InvalidateAllTransactions()
...
 
             var invalidTransactions = invalidItems.Select(p => p.Tx).ToArray();
-            foreach (IMemoryPoolTxObserverPlugin plugin in Plugin.TxObserverPlugins)
-                plugin.TransactionsRemoved(_system, MemoryPoolTxRemovalReason.NoLongerValid, invalidTransactions);
+            foreach (Transaction tx in invalidTransactions)
+                TransactionRemoved?.Invoke(this, new()
+                {
+                    Transaction = tx,
+                    Reason = TransactionRemovalReason.NoLongerValid
+                });
```
Comment (shargon) [inline src/neo/Ledger/MemoryPool.cs:488]: The same here

File: src/neo/Wallets/Wallet.cs [hunk 5]
```diff
@@ -779,15 +780,12 @@ public static Wallet Migrate(string path, string oldPath, string password, Proto
...
         }
 
-        private static IEnumerable<IWalletFactory> GetFactories()
+        public static void RegisterFactory(IWalletFactory factory)
         {
-            return Plugin.Plugins
-                .OfType<IWalletFactory>()
-                .Append(NEP6WalletFactory.Instance)
-                .Append(SQLite.SQLiteWalletFactory.Instance);
+            factories.Add(factory);
```
Comment (shargon) [inline src/neo/Wallets/Wallet.cs:788]: Check if already exists?

File: src/neo/Ledger/TransactionRemovedEventArgs.cs [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+// modifications are permitted.
+
+using Neo.Network.P2P.Payloads;
+
+namespace Neo.Ledger;
+
+public sealed class TransactionRemovedEventArgs
+{
+    public Transaction Transaction { get; init; }
+    public TransactionRemovalReason Reason { get; init; }
```
Comment (shargon) [inline src/neo/Ledger/TransactionRemovedEventArgs.cs:18]: Add summary?

**Comment by shargon** [comment]
I like this change :)

File: src/neo/Utility.cs (hunk 1 above)
Comment (erikzhang) [reply src/neo/Utility.cs:33]: We usually don't add `on` before event name.

File: src/neo/NeoSystem.cs (hunk 2 above)
Comment (erikzhang) [reply src/neo/NeoSystem.cs:122]: What do you mean?

File: src/neo/Ledger/MemoryPool.cs (hunk 3 above)
Comment (erikzhang) [reply src/neo/Ledger/MemoryPool.cs:318]: But `TransactionAdded` doesn't use array.

File: src/neo/Wallets/Wallet.cs (hunk 5 above)
Comment (erikzhang) [reply src/neo/Wallets/Wallet.cs:788]: Why a plugin add itself twice?

File: src/neo/Ledger/TransactionRemovedEventArgs.cs (hunk 6 above)
Comment (erikzhang) [reply src/neo/Ledger/TransactionRemovedEventArgs.cs]: Done.

File: src/neo/Ledger/MemoryPool.cs (hunk 3 above)
Comment (shargon) [reply src/neo/Ledger/MemoryPool.cs:318]: I know, but it's more optimal to use an array for remove.

File: src/neo/NeoSystem.cs (hunk 2 above)
Comment (erikzhang) [reply src/neo/NeoSystem.cs:122]: The previous logic didn't do that. If do it, it should be done in neo-node.

File: src/neo/Ledger/MemoryPool.cs (hunk 3 above)
Comment (erikzhang) [reply src/neo/Ledger/MemoryPool.cs]: Done.

File: src/neo/Ledger/MemoryPool.cs [hunk 7]
```diff
@@ -27,6 +26,9 @@ namespace Neo.Ledger
     /// </summary>
     public class MemoryPool : IReadOnlyCollection<Transaction>
     {
+        public event EventHandler<Transaction> TransactionAdded;
+        public event EventHandler<TransactionRemovedEventArgs> TransactionRemoved;
```
Comment (shargon) [inline src/neo/Ledger/MemoryPool.cs:30]: Transaction**s**Removed?

**Pull request merged**